src/lang/processor.rs

```diff
@@ -0,0 +1,1181 @@
+                    });
+                }
+                first_request_span = Some(get_span(&pair));
+                ctx.request_pair = Some(pair);
+            }
+            Rule::EOI => break,
+            Rule::COMMENT | Rule::WHITESPACE => {}
+            _ => {
+                return Err(ProcessorError::RuleMismatch {
+                    expected_rule: Rule::custom_predicate_def,
+                    found_rule: pair.as_rule(),
+                    context: "top-level document content (expected custom_predicate_def, request_def, or EOI)".to_string(),
+                    span: Some(get_span(&pair)),
+                });
+            }
+        }
+    }
+    Ok(())
+}
+
+fn count_public_args(pred_def_pair: &Pair<Rule>) -> Result<usize, ProcessorError> {
+    let arg_section_pair = pred_def_pair
+        .clone()
+        .into_inner()
+        .find(|p| p.as_rule() == Rule::arg_section)
+        .unwrap();
+
+    let public_arg_list_pair = arg_section_pair
+        .into_inner()
+        .find(|p| p.as_rule() == Rule::public_arg_list)
+        .unwrap();
+
+    Ok(public_arg_list_pair
+        .into_inner()
+        .filter(|p| p.as_rule() == Rule::identifier)
+        .count())
+}
+
+fn second_pass(ctx: &mut ProcessingContext) -> Result<ProcessedOutput, ProcessorError> {
+    let mut cpb_builder =
+        CustomPredicateBatchBuilder::new(ctx.params.clone(), "PodlogBatch".to_string());
+
+    for pred_pair in &ctx.custom_predicate_pairs {
+        process_and_add_custom_predicate_to_batch(pred_pair, ctx, &mut cpb_builder)?;
+    }
+
+    let custom_batch = cpb_builder.finish();
+
+    let request_templates = if let Some(req_pair) = &ctx.request_pair {
+        process_request_def(req_pair, ctx, &custom_batch)?
+    } else {
+        Vec::new()
+    };
+
+    Ok(ProcessedOutput {
+        custom_batch,
+        request_templates,
+    })
+}
+
+fn pest_pair_to_builder_arg(
+    arg_content_pair: &Pair<Rule>,
+    context_stmt_name: &str,
+) -> Result<BuilderArg, ProcessorError> {
+    match arg_content_pair.as_rule() {
+        Rule::literal_value => {
+            let value = process_literal_value(arg_content_pair, context_stmt_name)?;
+            Ok(BuilderArg::Literal(value))
+        }
+        Rule::wildcard => {
+            let name = arg_content_pair.as_str().strip_prefix("?").unwrap();
+            Ok(BuilderArg::WildcardLiteral(name.to_string()))
+        }
+        Rule::anchored_key => {
+            let mut inner_ak_pairs = arg_content_pair.clone().into_inner();
+            let pod_id_pair = inner_ak_pairs.next().unwrap();
+
+            let pod_self_or_wc_str = match pod_id_pair.as_rule() {
+                Rule::wildcard => {
+                    let name = pod_id_pair.as_str().strip_prefix("?").unwrap();
+                    SelfOrWildcardStr::Wildcard(name.to_string())
+                }
+                Rule::self_keyword => SelfOrWildcardStr::SELF,
+                _ => {
```

Comment:
the `anchored_key` rule by grammar definition can only `wildcard` or `self_keyword` for the first pair, so this should be `unreachable`.